Encode 128 bits for IPC's destination identifier

chirags27 opened the pull request:
#### 096f18835da5d23891a76c17b4f010f057dcdee2
<pre>
Encode 128 bits for IPC&apos;s destination identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=248929">https://bugs.webkit.org/show_bug.cgi?id=248929</a>

Reviewed by Alex Christensen.

This will allow me to fit a FrameIdentifier which is 128 bits.

* Source/WTF/wtf/Int128.h:
(WTF::DefaultHash&lt;UInt128&gt;::hash):
(WTF::DefaultHash&lt;UInt128&gt;::equal):
(WTF::DefaultHash&lt;Int128&gt;::hash):
(WTF::DefaultHash&lt;Int128&gt;::equal):
(WTF::HashTraits&lt;UInt128&gt;::constructDeletedValue):
(WTF::HashTraits&lt;UInt128&gt;::isDeletedValue):
(WTF::HashTraits&lt;Int128&gt;::constructDeletedValue):
(WTF::HashTraits&lt;Int128&gt;::isDeletedValue):
(IPC::ArgumentCoder&lt;WTF::UInt128&gt;::encode):
(IPC::ArgumentCoder&lt;WTF::UInt128&gt;::decode):
(IPC::ArgumentCoder&lt;WTF::Int128&gt;::encode):
(IPC::ArgumentCoder&lt;WTF::Int128&gt;::decode):
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::makeObjectIdentifier):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didReceiveMessage):
(WebKit::NetworkProcess::didReceiveSyncMessage):
(WebKit::NetworkProcess::findPendingDownloadLocation):
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::WaitForMessageState::WaitForMessageState):
(IPC::Connection::SyncMessageState::dispatchMessages):
(IPC::Connection::addWorkQueueMessageReceiver):
(IPC::Connection::removeWorkQueueMessageReceiver):
(IPC::Connection::addMessageReceiver):
(IPC::Connection::removeMessageReceiver):
(IPC::Connection::createSyncMessageEncoder):
(IPC::Connection::waitForMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithAsyncReply):
(IPC::Connection::send):
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::Connection::waitForMessageForTesting):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::Decoder):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::destinationID const):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::Encoder):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/IPC/MessageReceiverMap.cpp:
(IPC::MessageReceiverMap::addMessageReceiver):
(IPC::MessageReceiverMap::removeMessageReceiver):
* Source/WebKit/Platform/IPC/MessageReceiverMap.h:
* Source/WebKit/Platform/IPC/MessageSender.h:
(IPC::MessageSender::send):
(IPC::MessageSender::sendSync):
(IPC::MessageSender::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/ReceiverMatcher.h:
(IPC::ReceiverMatcher::ReceiverMatcher):
(IPC::ReceiverMatcher::createWithZeroAsAnyDestination):
(IPC::ReceiverMatcher::matches const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::startReceivingMessages):
(IPC::StreamServerConnection::stopReceivingMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_handler):
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::addMessageReceiver):
(WebKit::AuxiliaryProcess::removeMessageReceiver):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::addMessageReceiver):
(WebKit::AuxiliaryProcessProxy::removeMessageReceiver):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::send):
(WebKit::AuxiliaryProcessProxy::sendSync):
(WebKit::AuxiliaryProcessProxy::sendWithAsyncReply):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::addMessageReceiver):
(WebKit::WebProcessPool::removeMessageReceiver):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::evaluateJavaScriptCallback):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage):
(WebKit::IPCTestingAPI::JSMessageListener::jsDescriptionFromDecoder):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::didReceiveMessage):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCTestingAPI.mm:

Canonical link: <a href="https://commits.webkit.org/258484@main">https://commits.webkit.org/258484@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f6c840c7addfa1bcf731157eba8072cbeee84c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35162 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111418 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171594 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2149 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109154 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107873 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92622 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37162 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/105620 "Found 1 webkitpy python3 test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_receiver") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91224 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78887 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92445 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4795 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25527 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88656 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2421 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4904 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1968 "Found 1 new test failure: storage/indexeddb/request-with-null-open-db-request.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29479 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10961 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45023 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91575 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6652 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20473 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3069 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -191,6 +194,8 @@ class alignas(16) UInt128Impl {
   // Returns the highest value for a 128-bit unsigned integer.
   friend constexpr UInt128Impl UInt128Max();
 
+  bool isHashTableDeletedValue() const { return lo_ == std::numeric_limits<uint64_t>::max() && hi_ == std::numeric_limits<uint64_t>::max(); }
```
`UInt128Impl` should look like and behaves as if `uint128_t` because we are using `using UInt128 = UInt128Impl`. Thus, we should not have any methods.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -396,6 +401,8 @@ class alignas(16) Int128Impl {
   // Returns the minimum value for a 128-bit signed integer.
   friend constexpr Int128Impl Int128Min();
 
+  bool isHashTableDeletedValue() const { return lo_ == std::numeric_limits<uint64_t>::max() && hi_ == std::numeric_limits<int64_t>::max(); }
```
Ditto. We should not have method.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1271,72 @@ using UInt128 = UInt128Impl;
+#if !HAVE(INT128_T)
+template<> struct HashTraits<UInt128> : public SimpleClassHashTraits<UInt128> { };
+template<> struct HashTraits<Int128> : public SimpleClassHashTraits<Int128> { };
+#endif
```
We should make them based on `GenericHashTraits` since we should not treat UInt128 / Int128 as a class.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1271,72 @@ using UInt128 = UInt128Impl;
+    template<typename Encoder> static void encode(Encoder& encoder, const WTF::UInt128& i)
+    {
+        encoder << WTF::UInt128High64(i);
+        encoder << WTF::UInt128Low64(i);
```
These functions should not be used. This is implementation.
Do like,
```
encoder << static_cast<uint64_t>(i >> 64);
encoder << static_cast<uint64_t>(i);
```

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1271,72 @@ using UInt128 = UInt128Impl;
+        if (!low)
+            return std::nullopt;
+
+        return WTF::MakeUInt128(*high, *low);
```
Ditto. This function should not be used. Do usual integer things.
```
(static_cast<UInt128>(*high) << 64) | *low
```

chirags27 replied:
This is for a platform with `!HAVE(INT128_T)`, so `static_cast<UInt128>` won't work right?

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1271,72 @@ using UInt128 = UInt128Impl;
+    template<typename Encoder> static void encode(Encoder& encoder, const WTF::Int128& i)
+    {
+        encoder << WTF::Int128High64(i);
+        encoder << WTF::Int128Low64(i);
```
Ditto.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1271,72 @@ using UInt128 = UInt128Impl;
+        if (!low)
+            return std::nullopt;
+
+        return WTF::MakeInt128(*high, *low);
```
Ditto.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1267,73 @@ using UInt128 = UInt128Impl;
+#if !HAVE(INT128_T)
+template<> struct HashTraits<UInt128> : GenericHashTraits<UInt128> { };
+template<> struct HashTraits<Int128> : GenericHashTraits<Int128> { };
+#endif
```
How about just always defining this regardless of whether we are using uint128_t?
```
template<> struct HashTraits<UInt128> : GenericHashTraits<UInt128> {
    static constexpr bool emptyValueIsZero = true;
    static void constructDeletedValue(UInt128& slot) { slot = static_cast<UInt128 >(-1); }
    static bool isDeletedValue(UInt128 value) { return value == static_cast<UInt128 >(-1); }
};
```
And ditto for `Int128`.

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1267,73 @@ using UInt128 = UInt128Impl;
 
 }  // namespace WTF
 
+#if !HAVE(INT128_T)
```
Why? We can always define the following for `UInt128` regardless.
And I think this ifdefs are the reason why GCC is failing.

Constellation commented on Source/WTF/wtf/ObjectIdentifier.h:
```diff
@@ -132,6 +133,9 @@ template<typename T> class ObjectIdentifier : private ObjectIdentifierBase {
 
 private:
     template<typename U> friend ObjectIdentifier<U> makeObjectIdentifier(uint64_t);
+#if !HAVE(INT128_T)
```
Why? I think it should be defined always.

Constellation commented on Source/WTF/wtf/ObjectIdentifier.h:
```diff
@@ -152,6 +156,14 @@ template<typename T> inline ObjectIdentifier<T> makeObjectIdentifier(uint64_t id
     return ObjectIdentifier<T> { identifier };
 }
 
+#if !HAVE(INT128_T)
```
Why?

Constellation commented on Source/WTF/wtf/Int128.h:
```diff
@@ -1264,10 +1267,73 @@ using UInt128 = UInt128Impl;
+    static constexpr bool safeToCompareToEmptyOrDeleted = true;
+};
+
+#if !HAVE(INT128_T)
```
Why?

Constellation reviewed (changes requested):
Using `HAVE(INT128_T)` is wrong. It is very implementation detail thing.
Unless writing a code implementing `UInt128` itself, code touching `HAVE(INT128_T)` is probably not correct.

chirags27 committed:
Encode 128 bits for IPC's destination identifier
https://bugs.webkit.org/show_bug.cgi?id=248929

Reviewed by Alex Christensen.

This will allow me to fit a FrameIdentifier which is 128 bits.

* Source/WTF/wtf/Int128.h:
(WTF::DefaultHash<UInt128>::hash):
(WTF::DefaultHash<UInt128>::equal):
(WTF::DefaultHash<Int128>::hash):
(WTF::DefaultHash<Int128>::equal):
(WTF::HashTraits<UInt128>::constructDeletedValue):
(WTF::HashTraits<UInt128>::isDeletedValue):
(WTF::HashTraits<Int128>::constructDeletedValue):
(WTF::HashTraits<Int128>::isDeletedValue):
(IPC::ArgumentCoder<WTF::UInt128>::encode):
(IPC::ArgumentCoder<WTF::UInt128>::decode):
(IPC::ArgumentCoder<WTF::Int128>::encode):
(IPC::ArgumentCoder<WTF::Int128>::decode):
* Source/WTF/wtf/ObjectIdentifier.h:
(WTF::makeObjectIdentifier):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didReceiveMessage):
(WebKit::NetworkProcess::didReceiveSyncMessage):
(WebKit::NetworkProcess::findPendingDownloadLocation):
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::WaitForMessageState::WaitForMessageState):
(IPC::Connection::SyncMessageState::dispatchMessages):
(IPC::Connection::addWorkQueueMessageReceiver):
(IPC::Connection::removeWorkQueueMessageReceiver):
(IPC::Connection::addMessageReceiver):
(IPC::Connection::removeMessageReceiver):
(IPC::Connection::createSyncMessageEncoder):
(IPC::Connection::waitForMessage):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::sendWithAsyncReply):
(IPC::Connection::send):
(IPC::Connection::sendSync):
(IPC::Connection::waitForAndDispatchImmediately):
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::Connection::waitForMessageForTesting):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::Decoder::Decoder):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::destinationID const):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::Encoder::Encoder):
* Source/WebKit/Platform/IPC/Encoder.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/IPC/MessageReceiverMap.cpp:
(IPC::MessageReceiverMap::addMessageReceiver):
(IPC::MessageReceiverMap::removeMessageReceiver):
* Source/WebKit/Platform/IPC/MessageReceiverMap.h:
* Source/WebKit/Platform/IPC/MessageSender.h:
(IPC::MessageSender::send):
(IPC::MessageSender::sendSync):
(IPC::MessageSender::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/ReceiverMatcher.h:
(IPC::ReceiverMatcher::ReceiverMatcher):
(IPC::ReceiverMatcher::createWithZeroAsAnyDestination):
(IPC::ReceiverMatcher::matches const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::startReceivingMessages):
(IPC::StreamServerConnection::stopReceivingMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_handler):
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::addMessageReceiver):
(WebKit::AuxiliaryProcess::removeMessageReceiver):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::addMessageReceiver):
(WebKit::AuxiliaryProcessProxy::removeMessageReceiver):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::send):
(WebKit::AuxiliaryProcessProxy::sendSync):
(WebKit::AuxiliaryProcessProxy::sendWithAsyncReply):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::addMessageReceiver):
(WebKit::WebProcessPool::removeMessageReceiver):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::evaluateJavaScriptCallback):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPCStreamClientConnection::prepareToSendOutOfStreamMessage):
(WebKit::IPCTestingAPI::JSMessageListener::jsDescriptionFromDecoder):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::didReceiveMessage):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCTestingAPI.mm:

Canonical link: https://commits.webkit.org/258484@main
